File: 210-course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<Integer> ans = new ArrayList();
        // Build adjacency list
        List<List<Integer>> adjList = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            adjList.add(new ArrayList<>());
        }
        for (int[] course : prerequisites) {
            adjList.get(course[0]).add(course[1]);
        }

        // States: 0 = unvisited, 1 = visiting, 2 = visited
        int[] states = new int[numCourses];

        for (int i = 0; i < numCourses; i++) {
            if (!dfs(i, adjList, states, ans)) {
                return new int[]{};
            }
        }
        return ans.stream().mapToInt(i -> i).toArray();
    }

    private boolean dfs(int node, List<List<Integer>> adjList, int[] states, List<Integer> ans) {
        if (states[node] == 2) {
            return true; // already processed
        }
        if (states[node] == 1) {
            return false; // cycle detected
        }

        states[node] = 1; // mark as visiting

        for (int neighbor : adjList.get(node)) {
            if (!dfs(neighbor, adjList, states, ans)) {
                return false;
            }
        }

        states[node] = 2; // mark as visited
        ans.add(node);
        return true;
    }
}

File: 210-course-schedule-ii/course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adjList=defaultdict(list)

        ans=[]
        courses=prerequisites
        
        for course in courses:
            adjList[course[0]].append(course[1])
            
        UNVISITED=0
        VISITING=1
        VISITED=2
            
        states=[0]*numCourses
        
        def dfs(node):
            state=states[node]
            
            if state==VISITED:
                return True
            
            elif state==VISITING:
                return False
            
            states[node]=VISITING
            
            for nei in adjList[node]:
                if not dfs(nei):
                    return False
            
            states[node]=VISITED
            ans.append(node)
            return True
        
        for i in range(numCourses):
            if not dfs(i):
                return []
        
        return ans